source/validate_literals.cpp

```diff
@@ -0,0 +1,89 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Validates literal numbers.
+
+#include "validate.h"
+
+#include <cassert>
+
+#include "diagnostic.h"
+#include "opcode.h"
+#include "val/instruction.h"
+#include "val/validation_state.h"
+
+namespace libspirv {
+
+namespace {
+
+// Returns true if the operand holds a literal number
+bool IsLiteralNumber(const spv_parsed_operand_t*  operand) {
+  switch (operand->number_kind) {
+    case SPV_NUMBER_SIGNED_INT:
+    case SPV_NUMBER_UNSIGNED_INT:
+    case SPV_NUMBER_FLOATING:
+      return true;
+    default: return false;
+  }
+}
+
+// Verifies that the upper bits of the given upper |word| with given
+// lower |width| are zero- or sign-extended when |sign| is true
+bool verifyUpperBits(uint32_t word, uint32_t width, bool sign) {
+  assert(width < 32);
+  uint32_t upper_mask = (0xFFFFFFFF >> width) << width;
+  uint32_t upper_bits = word & upper_mask;
+  if (upper_bits == 0x00000000)
+      return true; // zero-extended
+  if (sign && upper_bits == upper_mask)
```

Comment:
The "sign" variable only says whether it's a signed number format.  It doesn't say whether the sign bit is 0 or 1.